ext/uri/php_uri.c

```diff
@@ -40,12 +48,565 @@ static const zend_module_dep uri_deps[] = {
+		RETURN_THROWS();
+	}
+	zval tmp;
+	ZVAL_STR(&tmp, uri_str);
+
+	array_init(return_value);
+
+	zval arr;
+	array_init(&arr);
+	zend_hash_str_add_new(Z_ARRVAL(arr), URI_SERIALIZED_PROPERTY_NAME, sizeof(URI_SERIALIZED_PROPERTY_NAME) - 1, &tmp);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &arr);
+
+	/* Serialize regular properties: second array */
+	ZVAL_ARR(&arr, this_object->handlers->get_properties(this_object));
+	Z_TRY_ADDREF(arr);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &arr);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __unserialize)
+{
+	uri_unserialize(INTERNAL_FUNCTION_PARAM_PASSTHRU, URI_PARSER_WHATWG);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __debugInfo)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *object = Z_OBJ_P(ZEND_THIS);
+
+	RETURN_ARR(uri_get_debug_properties(object));
+}
+
+static zend_object *uri_create_object_handler(zend_class_entry *class_type)
+{
+	uri_object_t *uri_object = zend_object_alloc(sizeof(uri_object_t), class_type);
+
+	uri_object->internal.uri = NULL;
+	uri_object->internal.handler = NULL;
+
+	zend_object_std_init(&uri_object->std, class_type);
+	object_properties_init(&uri_object->std, class_type);
+
+	return &uri_object->std;
+}
+
+static void uri_free_obj_handler(zend_object *object)
+{
+	uri_object_t *uri_object = uri_object_from_obj(object);
+	ZEND_ASSERT(uri_object != NULL);
+
+	if (UNEXPECTED(uri_object->internal.uri != NULL)) {
+		uri_object->internal.handler->free_uri(uri_object->internal.uri);
+		uri_object->internal.handler = NULL;
+		uri_object->internal.uri = NULL;
+	}
+
+	zend_object_std_dtor(&uri_object->std);
+}
+
+static zend_object *uri_clone_obj_handler(zend_object *object)
+{
+	uri_object_t *uri_object = uri_object_from_obj(object);
+	uri_internal_t *internal_uri = uri_internal_from_obj(object);
+
+	zend_object *new_object = uri_create_object_handler(object->ce);
+	ZEND_ASSERT(new_object != NULL);
+	uri_object_t *new_uri_object = uri_object_from_obj(new_object);
+
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	new_uri_object->internal.handler = internal_uri->handler;
+
+	void *uri = internal_uri->handler->clone_uri(internal_uri->uri);
+	ZEND_ASSERT(uri != NULL);
+
+	new_uri_object->internal.uri = uri;
+
+	zend_objects_clone_members(&new_uri_object->std, &uri_object->std);
+
+	return &new_uri_object->std;
+}
+
+PHPAPI void php_uri_implementation_set_object_handlers(zend_class_entry *ce, zend_object_handlers *object_handlers)
+{
+	ce->create_object = uri_create_object_handler;
+	ce->default_object_handlers = object_handlers;
+	memcpy(object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	object_handlers->offset = XtOffsetOf(uri_object_t, std);
+	object_handlers->free_obj = uri_free_obj_handler;
+	object_handlers->clone_obj = uri_clone_obj_handler;
+}
+
+zend_result uri_handler_register(const uri_handler_t *uri_handler)
+{
+	zend_string *key = zend_string_init_interned(uri_handler->name, strlen(uri_handler->name), 1);
```

Comment:
In some extreme edge cases, it's possible that this doesn't return an interned string. So you need to release it after the call to zend_hash_add_ptr. Furthermore if zend_hash_add_ptr were to return FAILURE you would also need to release it anyway.